jsettlers.logic/src/jsettlers/logic/map/grid/MainGrid.java

```diff
@@ -853,12 +885,16 @@ public final short getHeight() {
 			return height;
 		}
 
-		boolean canConstructAt(short x, short y, EBuildingType type, byte playerId) {
+		@Override
+		public boolean canConstructAt(short x, short y, EBuildingType type, byte playerId) {
 			ELandscapeType[] landscapes = type.getGroundtypes();
 			RelativePoint[] protectedTiles = type.getProtectedTiles();
 
 			int firstPosX = protectedTiles[0].calculateX(x);
 			int firstPosY = protectedTiles[0].calculateY(y);
+			if (firstPosX < 0 || firstPosY < 0 || firstPosX > partitionsGrid.getWidth() || firstPosY > partitionsGrid.getHeight()) {
```

Comment:
I changed the code as following:
```if (!isInBounds(firstPosX, firstPosY)) {
				return false;
			}
			final short partitionId = getPartitionIdAt(firstPosX, firstPosY);
```
I only check the first x,y pair because in the next line it would lead to an ArrayOutOfBounds Exception when the partitionID is requested.

Some lines below all relative positions were checked against inBounds by calling canUsePositionForConstruction

```
for (RelativePoint curr : protectedTiles) {
				int currX = curr.calculateX(x);
				int currY = curr.calculateY(y);

				if (!canUsePositionForConstruction(currX, currY, landscapes, partitionId)) {
					return false;
				}
			}
```

```
public boolean canUsePositionForConstruction(int x, int y, ELandscapeType[] landscapeTypes, short partitionId) {
			return isInBounds(x, y)
					&& !flagsGrid.isProtected(x, y)
					&& partitionsGrid.getPartitionIdAt(x, y) == partitionId
					&& isAllowedLandscape(x, y, landscapeTypes);
		}
```